### PR TITLE
[api][webui] Update Notification to use JSON serialisation.

### PR DESCRIPTION
**Pull request by evanrolfe**
Otherwise the Event.payload gets parsed from json, then serialised as YAML
into Notification.event_payload which sometimes causes errors because
the payload can fit into the events table (with JSON serialisation) but
might not fit into notifications.event_payload becuase YAML
serialisation takes up more characters than JSON does.

Here is the email I wrote to the open-build-service mailing list about using data_migrate gem:

> Hi all, I would like to propose that we use the data_migrate gem to handle changes to db data. Here is a PR
> I have opened where I need to update each notification's event_payload from YAML to JSON serialisation. This
> doesn't make sense to do in a normal migration because it has nothing to do with database structure only the
> content of the database. This gem allows us to handle such changes in the same way we handle normal
> migrations.
> 
> The alternative is to use rake tasks for data changes which is what I initially did in the PR in this commit.
> 
> However the downsides of that are that we need to always update the README's to let updaters know exactly
> which rake tasks need to be run to update their db and it also makes it easier to sync data changes with
> other developers. All thats needed to get your database up to date with the gem is to run `rake db:migrate:with_data`.
> 
> If you have any opinions please post on the PR to continue this discussion.
> 
> Thanks

### CODE REVIEW

File: src/api/db/migrate/20170829080916_convert_notifications_event_payload_to_json.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
+class ConvertNotificationsEventPayloadToJson < ActiveRecord::Migration[5.1]
+  def up
+    Notification20170829080916.transaction do
+      Notification20170829080916.all.find_each do |notification|
+        json = yaml_to_json(notification.event_payload)
+        Rails.logger.error "Converting Notification ##{notification.id}"
```
Comment (bgeuken) [inline src/api/db/migrate/20170829080916_convert_notifications_event_payload_to_json.rb]: That's not an error, right?

Comment (evanrolfe) [reply src/api/db/migrate/20170829080916_convert_notifications_event_payload_to_json.rb]: true, I've removed this line.

**Comment by evanrolfe** [comment]
I've added DO NOT MERGE because I think this would be better in a rake task rather than a migration.

**Comment by evanrolfe** [comment]
This is ready for review now.

File: ReleaseNotes-2.9 [hunk 2]
```diff
@@ -23,6 +23,11 @@ OBS Appliance users who have setup their LVM can just replace
 their appliance image without data loss. The migration will
 happen automatically.
 
+Upgrading from previous versions:
+=================================
+
+ * Please run this rake task if you notifications table has any rows in it:
```
Comment (hennevogel) [inline ReleaseNotes-2.9]: *your*. Setzen 6, Herr Rolfe.

File: src/api/lib/tasks/databases.rake [hunk 3]
```diff
@@ -1,3 +1,5 @@
+require_relative '../../app/models/application_record'
```
Comment (hennevogel) [inline src/api/lib/tasks/databases.rake]: Huh?

**Comment by evanrolfe** [comment]
> Can we separate them even more? Like in single files, with dates, like DB migrations?

Definitely.

> Also this would be the first time we require our users to do something like this. Or is this not mandatory?

Its mandatory if you have any rows in the notifications table.. Otherwise the Notification model won't be able to parse the event_payload of existing notifications.

File: src/api/lib/tasks/databases.rake (hunk 3 above)
Comment (evanrolfe) [reply src/api/lib/tasks/databases.rake]: If I dont include this then I get `uninitialized constant ApplicationRecord` when I run the rake task from the command line. I don't really understand it either..

**Comment by hennevogel** [comment]
> Its mandatory 

Okay then we even have to thing about more fuss about this. Hm....

**Comment by evanrolfe** [comment]
@hennevogel I've also used this gem before and it works really well. It basically treats data migrations just like normal migrations so you can just run `rake db:data_migrate`:

https://github.com/ilyakatz/data-migrate

**Comment by hennevogel** [comment]
Currently discussing this on our mailinglist. Stay tuned...

File: dist/README.UPDATERS [hunk 4]
```diff
@@ -195,7 +195,7 @@ use port 443 now.
  2) Restart memcached after database migration to wipe possible broken
     cache data.
 
- 3) configuration settings moved from various palces to unique /configuration 
+ 3) configuration settings moved from various palces to unique /configuration
```
Comment (bgeuken) [inline dist/README.UPDATERS]: It was already there, but...

*places

File: src/api/db/data/20170831143534_convert_notifications_event_payload_to_json.rb [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  end
+
+  private
+
+  def traverse_value(value, &block)
+    if value.is_a? Hash
+      value.each { |key, sub_value| value[key] = traverse_value(sub_value, &block) }
+
+    elsif value.is_a? Array
+      value.map { |element| traverse_value(element, &block) }
```
Comment (bgeuken) [inline src/api/db/data/20170831143534_convert_notifications_event_payload_to_json.rb]: Doesn't really matter, but is it map for a reason?

File: src/api/spec/db/data/convert_notifications_event_payload_to_json_spec.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
+require 'rails_helper'
+require Rails.root.join('db/data/20170831143534_convert_notifications_event_payload_to_json.rb')
+
+RSpec.describe ConvertNotificationsEventPayloadToJson do
```
Comment (bgeuken) [inline src/api/spec/db/data/convert_notifications_event_payload_to_json_spec.rb]: Nice! That's our first migration spec, right?

**Comment by hennevogel** [comment]
I'm missing this in `lib/tasks/databases.rake`

```
  desc 'Migrate the database (options: VERSION=x, VERBOSE=false, SCOPE=blog)'
  task migrate: :environment do
    puts ''
    puts 'warning: db:migrate only migrates your database structure, not the data contained in it.'
    puts 'warning for migrating your data run data:migrate'
    puts ''
  end
```

This will just append this warning to db:migrate

**Comment by evanrolfe** [comment]
I'm adding DO NOT MERGE since I haven't looked at this PR in some weeks so I want to review this again when I have time.

**Comment by hennevogel** [comment]
@evanrolfe are you ever going to have time for this? 🤠

**Comment by evanrolfe** [comment]
I want to do this through the sprint process so that we're all on board with this change. I created a card for it last week so I will bring this up in the planning today!

https://trello.com/c/0i5pBPkV/1003-change-notification-model-to-use-json-serialisation

File: src/api/spec/db/data/convert_notifications_event_payload_to_json_spec.rb (hunk 6 above)
Comment (evanrolfe) [reply src/api/spec/db/data/convert_notifications_event_payload_to_json_spec.rb]: I think so!

File: src/api/db/data/20170831143534_convert_notifications_event_payload_to_json.rb (hunk 5 above)
Comment (evanrolfe) [reply src/api/db/data/20170831143534_convert_notifications_event_payload_to_json.rb]: Yes, because this needs to return an array.

**Comment by evanrolfe** [comment]
@hennevogel this can be merged now!

**Pull request merged**